From bb58e3fda70d83f696f1183defc93e6501a69c55 Mon Sep 17 00:00:00 2001
From: Dmitry Kazakov <dimula73@gmail.com>
Date: Wed, 15 May 2019 19:39:44 +0300
Subject: [PATCH 1/2] Fix using tablet on QML widgets

In previous versions of Qt (wintab impeplementation) the events were
marked by Qt::MouseEventSynthesizedBySystem flag only when they were
synthesized from touch, not from tablet events. This is what
QWindowsTabletSupport does and what QQuickWindow expects (it
filters out all synthesized events). This patch recovers the old behavior
for the new QWindowsPointerHandler tablet API implementation.

See bug: https://bugs.kde.org/show_bug.cgi?id=406668
---
 src/plugins/platforms/windows/qwindowspointerhandler.cpp | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/plugins/platforms/windows/qwindowspointerhandler.cpp b/src/plugins/platforms/windows/qwindowspointerhandler.cpp
index 190fb208..07f4d41e 100644
--- a/src/plugins/platforms/windows/qwindowspointerhandler.cpp
+++ b/src/plugins/platforms/windows/qwindowspointerhandler.cpp
@@ -627,14 +627,16 @@ bool QWindowsPointerHandler::translatePenEvent(QWindow *window, HWND hwnd, QtWin
 #endif
 }
 
-static inline bool isMouseEventSynthesizedFromPenOrTouch()
+static inline bool isMouseEventSynthesizedFromPen()
 {
     // For details, see
     // https://docs.microsoft.com/en-us/windows/desktop/tablet/system-events-and-mouse-messages
     const LONG_PTR SIGNATURE_MASK = 0xFFFFFF00;
     const LONG_PTR MI_WP_SIGNATURE = 0xFF515700;
 
-    return ((::GetMessageExtraInfo() & SIGNATURE_MASK) == MI_WP_SIGNATURE);
+    const quint64 extraInfo = ::GetMessageExtraInfo();
+
+    return ((extraInfo & SIGNATURE_MASK) == MI_WP_SIGNATURE) && (extraInfo & 0x80);
 }
 
 bool QWindowsPointerHandler::translateMouseWheelEvent(QWindow *window,
@@ -705,7 +707,7 @@ bool QWindowsPointerHandler::translateMouseEvent(QWindow *window,
     }
 
     Qt::MouseEventSource source = Qt::MouseEventNotSynthesized;
-    if (isMouseEventSynthesizedFromPenOrTouch()) {
+    if (isMouseEventSynthesizedFromPen()) {
         if (QWindowsIntegration::instance()->options() & QWindowsIntegration::DontPassOsMouseEventsSynthesizedFromTouch)
             return false;
         source = Qt::MouseEventSynthesizedBySystem;
-- 
2.20.1.windows.1

